api/handlers.go

```diff
@@ -289,13 +303,42 @@ func (a *API) Unmount(w http.ResponseWriter, r *http.Request) {
 
 	volName := volConfig.String()
 
+	ut := &config.UseMount{
+		Volume:   volName,
+		Reason:   lock.ReasonMount,
+		Hostname: a.Hostname,
+	}
+
+	if !volConfig.Unlocked {
+		a.RemoveStopChan(volName)
+
+		// XXX to doubly ensure we do not UNMOUNT something that is held elsewhere
+		// (presumably because it is mounted THERE instead), we refuse to unmount
+		// anything that doesn't acquire a lock. We also remove the TTL refresh
+		// before taking it so it is not cleared in the unlikely event the mount
+		// takes longer than the TTL. Re-establish the TTL on error only if it is in
+		// locked mode.
+		if err := a.Client.PublishUse(ut); err != nil {
+			a.HTTPError(w, errors.LockFailed.Combine(err))
+			return
+		}
+	}
+
 	if a.MountCounter.Sub(volName) > 0 {
 		log.Warnf("Duplicate unmount of %q detected: ignoring and returning success", volName)
 		path, err := a.getMountPath(driver, driverOpts)
 		if err != nil {
 			a.HTTPError(w, errors.MarshalResponse.Combine(err))
 			return
 		}
+
+		if !volConfig.Unlocked {
```

Comment:
I can assure you it's executed in locked mode; everything would hang in the tests otherwise.

The great thing about the mount/unmount code is that it's easy to test in systemtests. nothing will work if it's broken.

The bad thing about the mount/unmount code is that it all requires multi-host functionality, and our UNIT tests are not sufficient to test this behavior.